rfc-117-containerised-runtime.md

```diff
@@ -0,0 +1,68 @@
+# RFC 000: Use a containerised runtime
+
+## Summary
+
+Use a containerised runtime for hosting GOV.UK applications. GOV.UK applications currently run on virtual machines running Ubuntu Trusty Tahr. There is currently an extended support contract for Ubuntu Trusty which will end in April 2022. Rather than upgrading Ubuntu on the virtual machines GOV.UK should change the application runtime to a containerised platform.
+
+## Problem
+
+GOV.UK currently hosts the majority of its applications and services using virtual machines running Ubuntu 14.04 LTS (Trusty Tahr).
+
+Trusty is out of date, it was released in April 2014 and standard support ended in April 2019. At this time GOV.UK signed up for [Extended Security Maintenance](https://ubuntu.com/esm) (ESM) which provides continued security fixes for high and critical common vulnerabilities and exposures (CVEs) for the [most commonly used packages](https://wiki.ubuntu.com/SecurityTeam/ESM/14.04#A14.04_Infrastructure_ESM_Packages). ESM will end in April 2022.
+
+Running an old operating system is preventing GOV.UK from upgrading key components of our technology stack (MongoDB, Puppet, Python, ...).
+
+Upgrading Ubuntu in the existing infrastructure is difficult for a number of reasons, including but not limited to:
+
+- Multiple applications share the operating system of the same virtual machine. This means it’s not easy to upgrade a single application at a time. Furthermore, the way our [Puppet code](http://github.com/alphagov/govuk-puppet/) is written makes supporting virtual machines on different operating systems at the same time hard to support.
+- GOV.UK currently uses upstart as our init system, but newer versions of Ubuntu switch to systemd. This means GOV.UK will need to do a significant rewrite of our init scripts as part of an upgrade.
+- Our current version of Puppet is 3.8, which is no longer supported (the current version is 6.13). Support for newer operating system versions is likely to be lacking in unsupported versions of Puppet.
+- Other infrastructure as code tools in use (e.g. [Fabric](https://github.com/alphagov/fabric-scripts)) are also currently using very old versions which are very likely to have problems with newer operating systems.
+
+## Proposal
+
+GOV.UK will adopt a container based hosting environment, such as [GOV.UK PaaS](https://www.cloud.service.gov.uk). The choice of hosting environment is out of scope for this RFC.
+
+GOV.UK applications will be built into container images, which will include the operating system. This may be done using buildpacks or other means - this decision is also out of scope for this RFC.
```

Comment:
```suggestion
GOV.UK applications and their dependencies will be built into container images. This may be done using buildpacks or other means - this decision is also out of scope for this RFC.
```
Not sure its clear or accurate to say they will include "the operating system".

Reply:
I also like this suggestion, as just containerizing the application and it's dependencies is far preferable in my opinion to starting with the state of a whole operating system and adding an application on top. If you're directly dealing with just your application and it's dependencies the state, and changes to it become much more explicit, which can make debugging easier, and avoid problems being introduced in unclear ways.

Using images for the container state is one decision, but if how to build those images is out of scope, then basing the images on operating systems should probably be out of scope as well.

Reply:
I agree - this wasn't meant to imply "we'll use `FROM ubuntu` or equivalent, not `FROM scratch`". I think that decision is out of scope for this RFC - the point is just we'll move away from having a shared operating system provided by the runtime.